src/game/puppyprint.c

```diff
@@ -878,148 +1182,269 @@ void print_small_text(s32 x, s32 y, const char *str, s32 align, s32 amount, s32
+
+        if (font != FONT_OUTLINE) {
+            font_offsets(offsetY, font, letter);
+        }
+    } else {
+        *textX = -12;
+        *spaceX = 2;
     }
+}
 
-    switch (letter) {
-        case '-': *textX = 40; *textY =  0; *spaceX = textLen[10]; break; // Hyphen
-        case '+': *textX = 44; *textY =  0; *spaceX = textLen[11]; break; // Plus
-        case '(': *textX = 48; *textY =  0; *spaceX = textLen[12]; break; // Open Bracket
-        case ')': *textX = 52; *textY =  0; *spaceX = textLen[13]; break; // Close Bracket
-        case '!': *textX = 56; *textY =  0; *spaceX = textLen[14]; break; // Exclamation mark
-        case '?': *textX = 60; *textY =  0; *spaceX = textLen[15]; break; // Question mark
-
-        case '"': *textX = 40; *textY = 12; *spaceX = textLen[42]; break; // Speech mark
-        case'\'': *textX = 44; *textY = 12; *spaceX = textLen[43]; break; // Apostrophe
-        case ':': *textX = 48; *textY = 12; *spaceX = textLen[44]; break; // Colon
-        case ';': *textX = 52; *textY = 12; *spaceX = textLen[45]; break; // Semicolon
-        case '.': *textX = 56; *textY = 12; *spaceX = textLen[46]; break; // Full stop
-        case ',': *textX = 60; *textY = 12; *spaceX = textLen[47]; break; // Comma
-
-        case '~': *textX = 40; *textY = 24; *spaceX = textLen[74]; break; // Tilde
-        case '@': *textX = 44; *textY = 24; *spaceX = textLen[75]; break; // Umlaut
-        case '^': *textX = 48; *textY = 24; *spaceX = textLen[76]; break; // Caret
-        case '/': *textX = 52; *textY = 24; *spaceX = textLen[77]; break; // Slash
-        case '_': *textX = 56; *textY = 24; *spaceX = textLen[78]; break; // Percent
-        case '&': *textX = 60; *textY = 24; *spaceX = textLen[79]; break; // Ampersand
-
-        // This is for the letters that sit differently on the line. It just moves them down a bit.
-        case 'g': *offsetY = 1; break;
-        case 'q': *offsetY = 1; break;
-        case 'p': if (font == FONT_DEFAULT) *offsetY = 3; break;
-        case 'y': if (font == FONT_DEFAULT) *offsetY = 1; break;
+u8 gIsLightText = FALSE;
+
+// This is where the deferred printing will be stored. When text is made, it will store text with an 12 byte header, then the rest will be the text data itself.
+// The first 4 bytes of the header will be the X and Y pos
+// The next 4 bytes will be the current envcolour set by print_set_envcolour
+// Then the string length, text alignment, amount and font each get a byte.
+// The data afterwards is the text data itself, using the string length byte to know when to stop.
+#define HEADERSIZE 13
+void print_small_text_buffered(s32 x, s32 y, const char *str, u8 align, s32 amount, u8 font) {
+    u8 strLen = MIN((signed)strlen(str), 255);
+    // Compare the cursor position and the string length, plus 12 (header size) and return if it overflows.
+    if (sPuppyprintTextBufferPos + strLen + HEADERSIZE > sizeof(sPuppyprintTextBuffer))
+        return;
+    x += 2048;
```

Comment:
Can these be calculated offsets rather than hardcoded jumps?

Reply:
I think I misunderstood what's going on here. Still seems weird but I'll allow it.